devicetypes/fibargroup/fibaro-roller-shutter.src/fibaro-roller-shutter.groovy

```diff
@@ -0,0 +1,728 @@
+			state "default", label:'${currentValue} W'
+		}
+		valueTile("energy", "device.energy", decoration: "flat", width: 2, height: 2) {
+			state "default", label:'${currentValue} kWh'
+		}
+		standardTile("refresh", "device.power", inactiveLabel: false, decoration: "flat", width: 2, height: 2) {
+			state "default", label:'', action:"refresh.refresh", icon:"st.secondary.refresh"
+		}
+		
+		details(["power", "energy", "refresh"])
+	}
+
+	preferences {
+		parameterMap.each {
+			input (
+					title: it.name,
+					description: it.description,
+					type: "paragraph",
+					element: "paragraph"
+			)
+
+			switch(it.type) {
+				case "boolRange":
+					input(
+							name: it.key + "Boolean",
+							type: "bool",
+							title: "Enable",
+							description: "If you disable this option, it will overwrite setting below.",
+							defaultValue: it.defaultValue != it.disableValue,
+							required: false
+					)
+					input(
+							name: it.key,
+							type: "number",
+							title: "Set value (range ${it.range})",
+							defaultValue: it.defaultValue,
+							range: it.range,
+							required: false
+					)
+				break
+				case "boolean":
+					input(
+							type: "paragraph",
+							element: "paragraph",
+							description: "Option enabled: ${it.activeDescription}\n" +
+									"Option disablePending: ${it.inactiveDescription}"
+					)
+					input(
+							name: it.key,
+							type: "boolean",
+							title: "Enable",
+							defaultValue: it.defaultValue == it.activeOption,
+							required: false
+					)
+				break
+				case "enum":
+					input(
+							name: it.key,
+							title: "Select",
+							type: "enum",
+							options: it.values,
+							defaultValue: it.defaultValue,
+							required: false
+					)
+				break
+				case "range":
+					input(
+							name: it.key,
+							type: "number",
+							title: "Set value (range ${it.range})",
+							defaultValue: it.defaultValue,
+							range: it.range,
+							required: false
+					)
+				break
+			}
+		}
+	}
+}
+
+def installed() {
+	state.calibrationStatus = "notStarted"
+	state.childDevices = [:]
+	state.childDevices.main = null
+	state.childDevices.assistant = null
+	state.currentMode = "windowShade"
+	createMainChildDeviceIfNeeded()
+	createAssistantChildDeviceIfNeeded()
+	sendEvent(name: "checkInterval", value: 2 * 60 * 60, displayed: false, data: [protocol: "zwave", hubHardwareId: device.hub.hardwareID])
+	// Preferences template begin
+	state.currentPreferencesState = [:]
+	parameterMap.each {
+		state.currentPreferencesState."$it.key" = [:]
+		state.currentPreferencesState."$it.key".value = getPreferenceValue(it)
+		if (it.type == "boolRange" && getPreferenceValue(it) == it.disableValue) {
+			state.currentPreferencesState."$it.key".status = "disablePending"
+		} else {
+			def preferenceName = it.key + "Boolean"
+			settings."$preferenceName" = true
+			state.currentPreferencesState."$it.key".status = "synced"
+		}
+	}
+	// Preferences template end
+}
+
+def updated() {
```

Comment:
For all of these preferences, I wonder if it's worth putting in code such that we don't set/query every parameter on install (or at least not default values)? I think it's safe to assume that devices have their factory default values on install.

Maybe another state when we initialize the preference map that denotes that it's still on default? I just worry about the storm of communication this requires.

Reply:
I'm not sure if I get you correctly, but right now ` it's safe to assume that devices have their factory default values on install` it's exactly what happens here. 
In `installed()` I'm setting all values of settings according to theirs defaults. There's no any Z-Wave communication at this point.
Then, if any parameter is updated by user, we're comparing `state.currentPreferencesState` and `settings` in search of differences. If there's any, DTH is sending configuration commands related just to those particular parameters. 

I'm not sure if I'm able to achieve lower radio traffic than that.